compiler/src/formatting/format.re

```diff
@@ -744,238 +760,252 @@ let rec block_item_iterator =
-             ~followed_by_arrow?,
-             remainder,
-           ),
-      ];
+  if (separated^) {
+    Doc.concat(items);
+  } else {
+    Doc.ifBreaks(
+      Doc.concat([separator, Doc.concat(items)]),
+      Doc.concat([Doc.concat(items), separator]),
+    );
+  };
+};
 
-    | [] =>
-      let comments_without_item_comments =
-        remove_used_comments(~remove_comments=item_comments, comments);
-      let trailing_comments =
-        Comment_utils.get_comments_after_location(
-          ~location=get_loc(item),
-          comments,
-        );
+let rec item_iterator =
+        (
+          ~previous: option(Location.t)=?,
+          ~get_loc: 'a => Location.t,
+          ~print_item: (~comments: list(Parsetree.comment), 'a) => Doc.t,
+          ~comments: list(Parsetree.comment),
+          ~separator,
+          ~followed_by_arrow: option(bool)=?,
+          ~iterated_item: iterator_item_type,
+          items: list('a),
+        ) => {
+  let trailing_separator =
+    switch (iterated_item) {
+    | IteratedListPattern
+    | IteratedDataDeclarations
+    | IteratedMatchItem
+    | IteratedRecordLabels => true
+    | IteratedRecordPattern
+    | IteratedRecord
+    | IteratedTypeConstructor
+    | IteratedPatterns
+    | IteratedTupleExpression
+    | IteratedArrayExpression
+    | IteratedTypeItems
+    | IteratedTupleConstructor
+    | IteratedEnum
+    | IteratedRecordData
+    | IteratedValueBindings => false
+    };
 
-      let bcb =
-        before_comments_break(
-          ~previous=PreviousItem(get_loc(item)),
-          ~this_line,
-          trailing_comments,
-        );
-      let this_item =
-        switch (leading_comments, trailing_comments) {
-        | ([leading, ..._], [trailing, ..._]) =>
-          let last_comment = get_last_item_in_list(leading_comments);
-
-          let filler =
-            switch (last_comment) {
-            | Line(_) => Doc.nil
-            | _ => Doc.space
-            };
-          Doc.concat([
-            Comment_utils.new_comments_to_docs(leading_comments),
-            filler,
-            print_item(~comments=item_comments, item),
-            switch (followed_by_arrow) {
-            | Some(true) =>
-              switch (trailing) {
-              | Line(cmt)
-              | Shebang(cmt) => Doc.concat([bcb, Doc.text(cmt.cmt_source)])
-              | _ =>
-                Doc.concat([
-                  bcb,
-                  Comment_utils.block_trailing_comments_docs(
-                    trailing_comments,
-                  ),
-                ])
-              }
-            | _ =>
-              Doc.lineSuffix(
-                Doc.concat([
-                  bcb,
-                  Comment_utils.block_trailing_comments_docs(
-                    trailing_comments,
-                  ),
-                ]),
-              )
-            },
-          ]);
+  let number_items = List.length(items);
+
+  if (number_items < 1) {
+    []; // nothing to do
+  } else {
```

Comment:
Can we turn this into a pattern match?

```
switch (items) {
  | [] => []
  | [first_item, ...rest] => { ... }
}
```
avoids the `List.length` and `List.hd` calls.

Reply:
Of course. I seem to have a blind spot for these!

Reply:
We need the list length lower though.